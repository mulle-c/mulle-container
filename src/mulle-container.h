//
//  mulle-container.h
//  mulle-container
//
//  Created by Nat! on 02/11/15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions are met:
//
//  Redistributions of source code must retain the above copyright notice, this
//  list of conditions and the following disclaimer.
//
//  Redistributions in binary form must reproduce the above copyright notice,
//  this list of conditions and the following disclaimer in the documentation
//  and/or other materials provided with the distribution.
//
//  Neither the name of Mulle kybernetiK nor the names of its contributors
//  may be used to endorse or promote products derived from this software
//  without specific prior written permission.
//
//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
//  AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
//  IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
//  ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
//  LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
//  CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
//  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
//  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
//  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
//  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
//  POSSIBILITY OF SUCH DAMAGE.
//
#ifndef mulle_container__h__
#define mulle_container__h__

#define MULLE_CONTAINER_VERSION  ((2 << 20) | (0 << 8) | 0)

#include "include.h"

#include "mulle-container-operation.h"
#include "mulle-container-math.h"
#include "mulle-container-callback.h"
#include "mulle-container-callback-global.h"

#include "mulle-prime.h"
#include "mulle-hash.h"

#include "mulle--array.h"
#include "mulle-array.h"

#include "mulle--set.h"
#include "mulle-set.h"

#include "mulle--map.h"
#include "mulle-map.h"

#include "mulle-range.h"
#include "mulle--rangeset.h"

#include "mulle--pointerqueue.h"
#include "mulle-pointerarray.h"
#include "mulle--pointermap.h"
#include "mulle-pointerset.h"
#include "mulle-pointerpairarray.h"

#include "mulle-container-callback.h"
#include "mulle-container-operation.h"

#if MULLE_C11_VERSION < ((3 << 20) | (2 << 8) | 0)
# error "mulle-c11 is too old"
#endif

#if MULLE_ALLOCATOR_VERSION < ((2 << 20) | (0 << 8) | 0)
# error "mulle-allocator is too old"
#endif


#endif /* mulle_container_h */
