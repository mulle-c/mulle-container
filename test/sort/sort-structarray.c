//
//  main.c
//  test-array
//
//  Created by Nat! on 04.11.15.
//  Copyright (c) 2015 Nat! - Mulle kybernetiK.
//  Copyright (c) 2015 Codeon GmbH.
//  All rights reserved.
//

#include <mulle-container/mulle-container.h>
#include <mulle-testallocator/mulle-testallocator.h>

#include <stdio.h>


struct demo_struct
{
   double   x;
   char     y;
};


int  reverse_compare( void *_a, void *_b, void *context)
{
   struct demo_struct  *a = _a;
   struct demo_struct  *b = _b;

   if( a->x > b->x)
      return( -1);
   return( a->x != b->x);
}


#define LOOPS   16


static void  simple( void)
{
   struct demo_struct                   a;
   struct demo_struct                   *item;
   struct mulle_structarray             *array;
   struct mulle_structarrayenumerator   rover;
   unsigned int                         i;

   array = mulle_structarray_create( sizeof( struct demo_struct),
                                     alignof( struct demo_struct),
                                     LOOPS,
                                     NULL);

   assert( mulle_structarray_get_last( array) == NULL);

   for( i = 0; i < LOOPS; i++)
   {
      a.x = (double) i + 1848;
      a.y = i + 'A';

      mulle_structarray_add( array, &a);
   }

   // more or less just a syntax check
   printf( "Initial:\n");
   i = 0;
   mulle_structarray_for( array, item)
   {
      assert( item);
      printf( "\t%u: %g=%c\n", i, item->x, item->y);
      ++i;
   }


   mulle_qsort_r( array->_storage, LOOPS, sizeof( struct demo_struct), reverse_compare, NULL);

   printf( "\nReversed:\n");
   // more or less just a syntax check
   i = 0;
   mulle_structarray_for( array, item)
   {
      assert( item);
      printf( "\t%u: %g=%c\n", i, item->x, item->y);
      ++i;
   }

   mulle_structarray_destroy( array);
}


// the mulle_testallocator detects and aborts on leaks
static void  run_test( void (*f)( void), char *name)
{
   mulle_testallocator_reset();
  // printf( "%s\n", name);
   (f)();
   mulle_testallocator_reset();
}


int main(int argc, const char * argv[])
{
   run_test( simple, "simple");
   return( 0);
}
