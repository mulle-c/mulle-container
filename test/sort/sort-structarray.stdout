Initial:
	0: 1848=A
	1: 1849=B
	2: 1850=C
	3: 1851=D
	4: 1852=E
	5: 1853=F
	6: 1854=G
	7: 1855=H
	8: 1856=I
	9: 1857=J
	10: 1858=K
	11: 1859=L
	12: 1860=M
	13: 1861=N
	14: 1862=O
	15: 1863=P

Reversed:
	0: 1863=P
	1: 1862=O
	2: 1861=N
	3: 1860=M
	4: 1859=L
	5: 1858=K
	6: 1857=J
	7: 1856=I
	8: 1855=H
	9: 1854=G
	10: 1853=F
	11: 1852=E
	12: 1851=D
	13: 1850=C
	14: 1849=B
	15: 1848=A
